gremlin-go/driver/gremlinlang.go

```diff
@@ -0,0 +1,502 @@
+	sb := strings.Builder{}
+	sb.WriteString("[")
+	size := reflect.ValueOf(arg).Len()
+	if size == 0 {
+		sb.WriteString(":")
+	} else {
+		iter := reflect.ValueOf(arg).MapRange()
+		for iter.Next() {
+			k := iter.Key().Interface()
+			kString, err := gl.argAsString(k)
+			if err != nil {
+				return "", err
+			}
+			v := iter.Value().Interface()
+			vString, err := gl.argAsString(v)
+			if err != nil {
+				return "", err
+			}
+			sb.WriteString(kString)
+			sb.WriteByte(':')
+			sb.WriteString(vString)
+			size--
+			if size > 0 {
+				sb.WriteString(",")
+			}
+		}
+	}
+
+	sb.WriteString("]")
+	return sb.String(), nil
+}
+
+func (gl *GremlinLang) translateSlice(arg interface{}) (string, error) {
+	sb := strings.Builder{}
+	sb.WriteString("[")
+	list := reflect.ValueOf(arg)
+
+	for i := 0; i < list.Len(); i++ {
+		if i > 0 {
+			sb.WriteString(",")
+		}
+		vString, err := gl.argAsString(list.Index(i).Interface())
+		if err != nil {
+			return "", err
+		}
+		sb.WriteString(vString)
+	}
+	sb.WriteString("]")
+	return sb.String(), nil
+}
+
+func (gl *GremlinLang) translateTextPredicate(v *textP) (string, error) {
+	if v.operator == "" || len(v.values) == 0 {
+		return "", nil
+	}
+
+	instructionString := ""
+	instructionString += v.operator
+	instructionString += "("
+	if len(v.values) == 1 {
+		argString, err := gl.argAsString(v.values[0])
+		if err != nil {
+			return "", err
+		}
+		instructionString += argString
+	} else if len(v.values) > 1 {
+		for index, arg := range v.values {
+			argString, err := gl.argAsString(arg)
+			if err != nil {
+				return "", err
+			}
+
+			instructionString += argString
+			if index < len(v.values)-1 && argString != "" {
+				instructionString += ","
+			}
+		}
+	}
+
+	instructionString += ")"
+
+	return instructionString, nil
+}
+
+func (gl *GremlinLang) translatePredicate(v *p) (string, error) {
+
+	if v.operator == "" || len(v.values) == 0 {
+		return "", nil
+	}
+
+	instructionString := ""
+	instructionString += v.operator
+	instructionString += "("
+	if len(v.values) == 1 {
+		argString, err := gl.argAsString(v.values[0])
+		if err != nil {
+			return "", err
+		}
+		instructionString += argString
+	} else if len(v.values) > 1 {
+		if v.operator != "between" && v.operator != "inside" {
+			instructionString += "["
+		}
+		for index, arg := range v.values {
+			argString, err := gl.argAsString(arg)
+			if err != nil {
+				return "", err
+			}
+
+			instructionString += argString
+			if index < len(v.values)-1 && argString != "" {
+				instructionString += ","
+			}
+		}
+		if v.operator != "between" && v.operator != "inside" {
+			instructionString += "]"
+		}
+	}
+
+	instructionString += ")"
+
+	return instructionString, nil
+}
+func (gl *GremlinLang) asParameter(arg interface{}) string {
+	paramName := fmt.Sprintf("_%d", gl.paramCount.Load())
+	gl.paramCount.Add(1)
+	gl.parameters[paramName] = arg
+	return paramName
+}
+
+func (gl *GremlinLang) GetGremlin(arg ...string) string {
```

Comment:
Nit: I wonder if we should limit using this vargargs pattern for psuedo-overloaded methods in favour of something more idiomatic in go. Feel free to ignore this request if you'd like as this pattern is already heavily used throughout gremlin-go.